[14.0][IMP] cetmix_tower_server: add reference to models

dessanhemrayev opened the pull request:
This commit adds references to the following models:

- cx.tower.plan.line
- cx.tower.plan.line.action
- cx.tower.variable.value

Changes

Inherit from cx.tower.reference.mixin
Add reference fields to views, including search view 
Update tests

Task: 4017

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Enhanced duplication process for flight plans and variable values, ensuring integrity during copying.
  - New `create` methods added to various models to manage references effectively during record creation.
  - New `reference` field added to the `cx.tower.plan.line`, `cx.tower.variable.value`, and `cx.tower.plan` views, improving data handling and visibility control based on user permissions.

- **Bug Fixes**
  - Improved test cases for plan execution and command handling, ensuring accurate validation.

- **Tests**
  - Expanded test coverage for reference preparation and population functionalities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_variable_value_view.xml:
```diff
@@ -1,11 +1,12 @@
         <field name="model">cx.tower.variable.value</field>
         <field name="arch" type="xml">
             <tree editable="top" decoration-bf="is_global==True">
+                <field name="name" />
```
This field has no sense for this model. So no need to show it

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_variable_value_view.xml:
```diff
@@ -1,11 +1,12 @@
         <field name="arch" type="xml">
             <tree editable="top" decoration-bf="is_global==True">
+                <field name="name" />
+                <field name="reference" optional="show" />
```
Should be hidden by default

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_variable_value_view.xml:
```diff
@@ -21,12 +22,13 @@
         <field name="model">cx.tower.variable.value</field>
         <field name="arch" type="xml">
             <search string="Search Values">
+                <field name="name" />
```
This field doesn't have any use for this model, remove it.

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_server_view.xml:
```diff
@@ -237,6 +237,7 @@
                             <field name="variable_value_ids" order="variable_reference">
                                 <tree editable="bottom">
                                     <field name="variable_reference" invisible="1" />
+                                    <field name="name" />
```
Remove this field, no-one will be creating names manually

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_plan_line_view.xml:
```diff
@@ -10,7 +10,11 @@
                         <group>
                             <field name="sequence" />
-
+                            <field name="name" />
```
This field is technical in this model, no need to show it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -10,12 +10,15 @@
     active = fields.Boolean(related="plan_id.active", readonly=True)
     sequence = fields.Integer(default=10)
-    name = fields.Char(related="command_id.name", readonly=True)
+    name = fields.Char(related="command_id.name", store=True)
```
Probably we can try to leave it as is (non-stored, readonly) and see if it works.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan.py:
```diff
@@ -182,6 +182,7 @@ def _get_next_action_values(self, command_log):
                                 "variable_id": variable.id,
                                 "value_char": variable_value.value_char,
                                 "server_id": server.id,
+                                "name": variable_value.value_char,
```
Why do we need it? And why `variable_value.value_char` is used for the "name" field value?

dessanhemrayev replied:
Thanks, I'll fix it

ivs-cetmix commented on cetmix_tower_server/models/cetmix_tower.py:
```diff
@@ -61,6 +61,7 @@ def server_set_variable_value(self, server_reference, variable_reference, value)
         else:
             self.env["cx.tower.variable.value"].create(
                 {
+                    "name": value,
```
This field should not be defined manually

ivs-cetmix reviewed (changes requested):
It's not clear how and why someone should provide "name" values for technical records, such as plan lines, plan line actions and variable values.
Basically "name" field value should be either autogenerated (bad) or used from related model. We need to check if this is possible to make "name" field non-stored in order to avoid useless data storage.

dessanhemrayev committed:
[14.0][IMP] cetmix_tower_server: add reference to models
This commit adds references to the following models:

cx.tower.plan.line
cx.tower.plan.line.action
cx.tower.variable.value
Changes

Inherit from cx.tower.reference.mixin
Add reference fields to views, including search view
Update tests

Task: 4017

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -10,6 +12,9 @@
 class CxTowerPlanLine(models.Model):
     _name = "cx.tower.plan.line"
+    _inherit = [
+        "cx.tower.reference.mixin",
```
Since this mixin has been added you also need to change the `_compute_variable_ids` method. please check how to use this method in the case when a mixin is connected: e.g. [here](https://github.com/cetmix/cetmix-tower/blob/fd29f987a08f3eb9ae3f4ce902f1cc969a6f082f/cetmix_tower_server/models/cx_tower_command.py#L123)

dessanhemrayev replied:
Thanks, I'll check it and fix

ivs-cetmix replied:
@

> Since this mixin has been added you also need to change the `_compute_variable_ids` method. please check how to use this method in the case when a mixin is connected: e.g. [here](https://github.com/cetmix/cetmix-tower/blob/fd29f987a08f3eb9ae3f4ce902f1cc969a6f082f/cetmix_tower_server/models/cx_tower_command.py#L123)

@GabbasovDinar Could you please elaborate why this should be done?

GabbasovDinar replied:
@ivs-cetmix we need to refactor this method because now this model inherits the methods of mixin

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -187,3 +192,10 @@ def _skip(self, server, plan_log_record, **kwargs):
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            current_datetime = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")
```
Please move this before the cycle starts

GabbasovDinar replied:
you can use datetime from field:
```python
fields.Datetime.now()
```

Also use Server datetime format (I recommend using this method):
```python
from odoo.tools.misc import DEFAULT_SERVER_DATETIME_FORMAT
```

If this format does not suit you, then define your format as a constant in constants.py: 
```python
PLAN_LINE_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S.%f"
```

dessanhemrayev replied:
> Please move this before the cycle starts

I need unique data to create a reference (the "name" field is used to create a reference)

ivs-cetmix replied:
> > Please move this before the cycle starts
> 
> I need unique data to create a reference (the "name" field is used to create a reference)

@dessanhemrayev  'name' field should'n be unique for an unique reference to be generated. This case is already handled in the [create](https://github.com/cetmix/cetmix-tower/blob/2711a9e613a0c3f7728cf176665e78f74ec8372a/cetmix_tower_server/models/cx_tower_reference_mixin.py#L137) method of the mixin.

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -9,10 +11,14 @@
     _name = "cx.tower.variable.value"
     _description = "Cetmix Tower Variable Values"
+    _inherit = [
+        "cx.tower.reference.mixin",
```
please, fix `_compute_variable_ids` method

ivs-cetmix replied:
@GabbasovDinar what's wrong with it?

GabbasovDinar replied:
@ivs-cetmix you need to compare these methods when the mixin is connected and when it is not. method uses the methods of the mixin, in one case by accessing the mixin model (because it is not connected) and in the second case using inherited methods, so, we need refactoring this method, because the mixin was connected with the model.

ivs-cetmix replied:
@GabbasovDinar are you sure that you are talking about the reference mixin here?

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_plan_line_action.py:
```diff
@@ -60,19 +59,53 @@ def _compute_name(self):
+                vals.get("value_char"),
+                vals.get("custom_exit_code"),
+            )
+            vals.update({"name": name if name else "Plan Line Action"})
```
@ivs-cetmix Should be a translatable value this default value ? ("Plan Line Action")

ivs-cetmix replied:
@

> @ivs-cetmix Should be a translatable value this default value ? ("Plan Line Action")

@GabbasovDinar nope

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -186,11 +186,14 @@ def _get_copied_name(self):
-            copy_name = _(
-                "%(name)s (copy %(number)s)", name=original_name, number=str(counter)
-            )
+        if not self.env.context.get("skip_copy_action"):
+            while self.search_count([("name", "=", copy_name)]) > 0:
+                counter += 1
+                copy_name = _(
+                    "%(name)s (copy %(number)s)",
+                    name=original_name,
+                    number=str(counter),
+                )
```
you can reduce code nesting, which will improve its readability:

```python
if self.env.context.get("skip_copy_action"):
    return copy_name

while .......
```

please, check it.

And please add comments as to why we should use this context and only then execute the loop

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -277,3 +283,10 @@ def _check_single_assignment(self):
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            current_datetime = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")
```
use fields.Datetime.now() and see my comments above

dessanhemrayev replied:
I tested this method, because it rounds the values ​​I can't use it
```
datetime.datetime.now()
2024-10-31 19:54:14.310060

fields.Datetime.now() 
2024-10-31 19:54:14.000000
```

GabbasovDinar replied:
why? add comments to code, I don't understand

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -187,3 +192,10 @@ def _skip(self, server, plan_log_record, **kwargs):
+    def create(self, vals_list):
+        for vals in vals_list:
+            current_datetime = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")
+            vals["name"] = f"Plan line {current_datetime}"
```
Why do we need to create a name for a plan line? It's not used anywhere

dessanhemrayev replied:
You are right, but the generation method depends on this field. To avoid possible problems, we are forced to add the value this way. Also, the reference must be unique, so we include the date in them.

ivs-cetmix replied:
I agree, however it's [related field](https://github.com/cetmix/cetmix-tower/blob/2711a9e613a0c3f7728cf176665e78f74ec8372a/cetmix_tower_server/models/cx_tower_plan_line.py#L22) 
Meanwhile [reference](https://github.com/cetmix/cetmix-tower/blob/2711a9e613a0c3f7728cf176665e78f74ec8372a/cetmix_tower_server/models/cx_tower_reference_mixin.py#L138) field is also used during the creation of the record. 
So I would suggest something like this:

```python
@api.model_create_multi
def create(self, vals_list):
     PlanModel = self.env["cx.tower.plan"]

     # Store the last line index per plan
     plan_line_index_dict = {}

     # Populate vals with references
     for vals in vals_list:
        plan_id = vals.get("plan_id")
        # Get stored line index for this plan and increment it
        line_index = plan_line_index_dict.get(plan_id, 1) + 1
        # Line reference = <plan_reference>_<line_index>
        vals.update({"reference": f"{PlanModel.browse(plan_id).reference}_{line_index}"})
        # Save updated line index
        plan_line_index_dict.update({plan_id: line_index})

    # Let main function do the job
    return super().create(vals_list)
```

@GabbasovDinar your thoughts on that?

GabbasovDinar replied:
if I understand the problem correctly, then

```python

from collections import defaultdict


@api.model_create_multi
def create(self, vals_list):
    PlanModel = self.env["cx.tower.plan"]

    # Extract all unique plan_ids from vals_list
    plan_ids = {vals.get("plan_id") for vals in vals_list if vals.get("plan_id")}

    # Fetch all plan references in a single query
    plans = PlanModel.browse(plan_ids)
    plan_refs = {plan.id: plan.reference for plan in plans}

    # Initialize line indices for each plan_id
    plan_line_index_dict = defaultdict(int)

    # Populate vals with references
    for vals in vals_list:
        plan_id = vals.get("plan_id")
        plan_line_index_dict[plan_id] += 1
        line_index = plan_line_index_dict[plan_id]
        vals["reference"] = f"{plan_refs[plan_id]}_{line_index}"

    # Delegate creation to the superclass method
    return super().create(vals_list)
```

dessanhemrayev replied:
@ivs-cetmix  @GabbasovDinar 
what about if plan_id not in vals ?
eg: [here](https://github.com/cetmix/cetmix-tower/blob/2711a9e613a0c3f7728cf176665e78f74ec8372a/cetmix_tower_server/tests/test_plan.py#L37)

dessanhemrayev replied:
and what about this [line](https://github.com/cetmix/cetmix-tower/blob/2711a9e613a0c3f7728cf176665e78f74ec8372a/cetmix_tower_server/models/cx_tower_reference_mixin.py#L135)?

ivs-cetmix replied:
> what about if plan_id not in vals ? eg: [here](https://github.com/cetmix/cetmix-tower/blob/2711a9e613a0c3f7728cf176665e78f74ec8372a/cetmix_tower_server/tests/test_plan.py#L37)

@dessanhemrayev thank you for pointing on this. In this case we can create some default placeholder reference. Eg "line_with_no_plan"

ivs-cetmix replied:
> and what about this [line](https://github.com/cetmix/cetmix-tower/blob/2711a9e613a0c3f7728cf176665e78f74ec8372a/cetmix_tower_server/models/cx_tower_reference_mixin.py#L135)?

@dessanhemrayev good point! I think we need to add a condition here. Eg `if "name" in vals: ...`

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line_action.py:
```diff
@@ -60,19 +59,53 @@ def _compute_name(self):
+
+            # Add custom exit code if action presumes it
+            if action == "ec":
+                action_string = "{} {}".format(action_string, custom_exit_code)
```
Please use f-string instead of 'format'

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -186,11 +186,14 @@ def _get_copied_name(self):
-            copy_name = _(
-                "%(name)s (copy %(number)s)", name=original_name, number=str(counter)
-            )
+        if not self.env.context.get("skip_copy_action"):
```
Why do we need it?

dessanhemrayev replied:
Because our "name" field is not saved, a problem arises when copying "plan line action". The essence of the problem is that the [method ](https://github.com/cetmix/cetmix-tower/blob/fd29f987a08f3eb9ae3f4ce902f1cc969a6f082f/cetmix_tower_server/models/cx_tower_reference_mixin.py#L190)that generates the value for the "reference" field

ivs-cetmix replied:
> Because our "name" field is not saved, a problem arises when copying "plan line action". The essence of the problem is that the [method ](https://github.com/cetmix/cetmix-tower/blob/fd29f987a08f3eb9ae3f4ce902f1cc969a6f082f/cetmix_tower_server/models/cx_tower_reference_mixin.py#L190)that generates the value for the "reference" field

In this case you should use more relevant name, because 'skip_copy' is too generic.
Eg 'reference_mixin_skip_copy'

dessanhemrayev replied:
Thanks, I'll fix it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -277,3 +283,10 @@ def _check_single_assignment(self):
+    def create(self, vals_list):
+        for vals in vals_list:
+            current_datetime = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")
+            vals["name"] = f"Variable value {current_datetime}"
```
Why do we need it? Are we using field 'name' somewhere?

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line_action.py:
```diff
@@ -60,19 +59,53 @@ def _compute_name(self):
                 rec.name = _("Wrong action")
+
+    @api.model
+    def _generate_action_name(self, condition, action, value_char, custom_exit_code):
```
Why do we need to make this a separate function?

ivs-cetmix replied:
@dessanhemrayev this is still pending

dessanhemrayev replied:
I'll remove this function

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line_action.py:
```diff
@@ -60,19 +59,53 @@ def _compute_name(self):
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            name = self._generate_action_name(
```
Same as with plan lines - we can generate 'referecne' instead of 'name'

ivs-cetmix replied:
@dessanhemrayev this is still pending

GabbasovDinar reviewed (changes requested):
please, check comments

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -187,3 +191,32 @@ def _skip(self, server, plan_log_record, **kwargs):
+                vals["reference"] = f"{plan_refs[plan_id]}_{line_index}"
+            else:
+                # Handle the case where plan_id is not present
+                tmp_line_index = self._get_current_microsecond()
```
We can simply use some static value instead. It will be automatically modified by [this](https://github.com/cetmix/cetmix-tower/blob/9e8617fcdb7aa7c1c5cbbdbf193d9e5caf208ee8/cetmix_tower_server/models/cx_tower_reference_mixin.py#L42) function if needed.
You are using similar method in variable values, so why are you using milliseconds based approach here?

dessanhemrayev replied:
Thanks
You are right, but when we run the tests, [this function](https://github.com/cetmix/cetmix-tower/blob/9e8617fcdb7aa7c1c5cbbdbf193d9e5caf208ee8/cetmix_tower_server/models/cx_tower_reference_mixin.py#L42) does not work correctly
yes, it checks for the presence of such a value in the table and shows that it is unique, but still, when creating a record, the sql_constraints is triggered and the creation of the record is blocked

and because of this I had to use such a function

ivs-cetmix replied:
@dessanhemrayev this means that either that function or tests should be fixed. Using milliseconds creates barely usable references and besides from the theory it doesn't give a 100% guarantee from the duplicates.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -186,10 +188,17 @@ def _get_copied_name(self):
         copy_name = _("%(name)s (copy)", name=original_name)
 
         counter = 1
+        if self.env.context.get("reference_mixin_skip_copy"):
```
You are skipping copy and returning copied name at the same time. Why?

dessanhemrayev replied:
I'll fix it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -187,3 +191,32 @@ def _skip(self, server, plan_log_record, **kwargs):
+        PlanModel = self.env["cx.tower.plan"]
+
+        # Extract all unique plan_ids from vals_list
+        plan_ids = {vals.get("plan_id") for vals in vals_list if vals.get("plan_id")}
```
Although sets a faster for search creating and updating them is slower than lists.  Here is some [Stackoverflow](https://stackoverflow.com/questions/7717011/which-is-faster-and-why-set-or-list) for that.
Probably `add x to list if x is not in list` will be faster. I would suggest profiling it to get the answer.

dessanhemrayev replied:
Thanks
I checked it link and tested code block
```python
import random
import timeit

# Generate a list of dictionaries with random values
vals_list = [{'key_name': random.randint(0, 1000)} for _ in range(10000)]
key_name = 'key_name'

def func1():
    line_ids = []
    for vals in vals_list:
        line_id = vals.get(key_name)
        if line_id and line_id not in line_ids:
            line_ids.append(line_id)

def func2():
    plan_ids = {vals.get(key_name) for vals in vals_list if vals.get(key_name)}

# Benchmark the two functions
time_func1 = timeit.timeit(func1, number=100)
time_func2 = timeit.timeit(func2, number=100)

print(f"Time taken by func1: {time_func1:.6f} seconds")
print(f"Time taken by func2: {time_func2:.6f} seconds")

```
I got these results
![image](https://github.com/user-attachments/assets/25cdf377-c08d-4c24-8721-8405cb4b8594)

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line_action.py:
```diff
@@ -76,3 +78,33 @@ def _compute_name(self):
+            else:
+                # Handle the case where line_id is not present
+                plan_line_index_dict["no_line"] += 1
+                line_index = plan_line_index_dict["no_line"]
```
Where is that "no_line" key added in dict?

dessanhemrayev replied:
if you are talking raise a KeyError
we use defaultdict and defaultdict never raises a KeyError

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -208,7 +214,14 @@ def copy(self, default=None):
+
+        # skip copy name because this function use in models
+        # where it field name non store
+        default["name"] = (
```
copy() function will use that original name automatically. No need to set it in code.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -277,3 +284,34 @@ def _check_single_assignment(self):
                 )
+
+    @api.model_create_multi
+    def create(self, vals_list):
```
Same issues as in all other `create` function implementations.

GabbasovDinar commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -277,3 +284,29 @@ def _check_single_assignment(self):
+                vals["reference"] = f"value_with_no_variable_{line_index}"
+
+        # Delegate creation to the superclass method
+        return super().create(vals_list)
```
this method has the same code and is used in three different places, please make one common method (maybe in cx.tower.reference.mixin model) and use it, it will make it easier to maintain the code in the future

dessanhemrayev replied:
Thanks, I'll fix it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,66 @@ def _get_id_by_reference(self, reference):
         return records and records[0].id
+
+    @api.model
+    def _prepare_references(self, model, key_name, vals_list):
```
This function must be covered with tests

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,66 @@ def _get_id_by_reference(self, reference):
+        return {line.id: line.reference for line in lines}
+
+    @api.model
+    def _populate_references(self, model_name, field_name, vals_list, suffix=""):
```
This function must be covered with tests

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -68,13 +69,11 @@ def _generate_or_fix_reference(self, reference_source):
 
         # If exclude same records from search results
-        if self:
+        if self and not self.env.context.get("reference_mixin_skip_self"):
```
Why do we need to skip it?

dessanhemrayev replied:
when we copy the value of a variable, we cannot verify the uniqueness, because this point is added to the domain of another condition

ivs-cetmix replied:
Sorry, don't understand this quite well. Could you please elaborate it further with some examples?

dessanhemrayev replied:
when we copy
```python
variable_value.copy({"plan_line_action_id": new_action.id})
```
we can't check its reference for uniqueness
because here (line 72) in self (old_record)
we got domain where 
`domain = [("id", "not in", self.ids), ("reference","=", reference)] `
and it sometimes incorrect

ivs-cetmix replied:
What does it mean "incorrect"?

dessanhemrayev replied:
Maybe I expressed myself incorrectly. What I meant was that if the search domain is like this, we will not be able to get a reliable result because the domain limits the choice of data.

ivs-cetmix replied:
Ok, understood

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -129,7 +128,8 @@ def create(self, vals_list):
             for vals in vals_list:
-                vals["name"] = vals["name"].strip()
+                if "name" in vals:
+                    vals["name"] = vals["name"].strip()
```
What if `name == False?`

ivs-cetmix reviewed (changes requested):
You should cover your code with tests right when you write it. Especially for such complex scenarios where TDD (test driven development) is preferred.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,71 @@ def _get_id_by_reference(self, reference):
+            CxModel = self.env[model]
+        except KeyError as err:
+            raise ValueError(
+                f"Model '{model}' does not exist. Please provide a valid model name."
```
Messages should be translated properly

dessanhemrayev replied:
Thanks, I'll fix it

ivs-cetmix commented on cetmix_tower_server/tests/common.py:
```diff
@@ -208,6 +208,18 @@ def setUp(self, *args, **kwargs):
             }
         )
 
+        self.plan_reference_mixin = self.Plan.create(
```
This method is used in one test only, so it should be moved to that test `setUp`

ivs-cetmix commented on cetmix_tower_server/tests/common.py:
```diff
@@ -208,6 +208,18 @@ def setUp(self, *args, **kwargs):
+            {"name": "Test Plan reference mixin", "note": "Test Note reference mixin"}
+        )
+
+        self.plan_line_reference_mixin = self.plan_line.create(
```
Same here

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,71 @@ def _get_id_by_reference(self, reference):
+    @api.model
+    def _populate_references(self, model_name, field_name, vals_list, suffix=""):
+        """
+        Common method to populate references based on a specific field.
```
I still cannot understand what does this function do. 
Description should be more specific. Eg `Create record references in the vals list used for 'create' method. Use references of the related model as a basis`.  Or something like that.

dessanhemrayev replied:
okay,I'll fix it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,71 @@ def _get_id_by_reference(self, reference):
+        Common method to populate references based on a specific field.
+
+        Args:
+            model_name (str): The model name to extract references from.
```
Maybe 'Related model name'? Or can this be not a related record either?

dessanhemrayev replied:
Thanks, I'll fix it

ivs-cetmix commented on cetmix_tower_server/tests/test_reference_mixin.py:
```diff
@@ -92,3 +92,109 @@ def test_search_by_reference(self):
         search_result = self.ServerTemplate.get_by_reference("not_much_template")
         self.assertEqual(len(search_result), 0, "Result should be empty")
+
+    def test_prepare_references_valid_input(self):
```
Which function exactly is tested here? Opt for using `test_<function_name>_..` naming pattern for better readability.

dessanhemrayev replied:
Thanks, I'll fix it

ivs-cetmix commented on cetmix_tower_server/tests/test_reference_mixin.py:
```diff
@@ -92,3 +92,109 @@ def test_search_by_reference(self):
+            "The reference should match the expected value.",
+        )
+
+    def test_invalid_model_name_raises_error(self):
```
Same regarding the test name here

ivs-cetmix commented on cetmix_tower_server/tests/test_reference_mixin.py:
```diff
@@ -92,3 +92,109 @@ def test_search_by_reference(self):
+            "The error message should indicate an invalid model name.",
+        )
+
+    def test_empty_vals_list_returns_empty_dict(self):
```
And here

ivs-cetmix commented on cetmix_tower_server/tests/test_reference_mixin.py:
```diff
@@ -92,3 +92,109 @@ def test_search_by_reference(self):
+            "The result should be an empty dictionary when vals_list is empty.",
+        )
+
+    def test_references_populated_correctly(self):
```
And here

ivs-cetmix commented on cetmix_tower_server/tests/test_reference_mixin.py:
```diff
@@ -92,3 +92,109 @@ def test_search_by_reference(self):
+            "The reference should be correctly populated with the suffix.",
+        )
+
+    def test_handling_missing_field_in_vals_list(self):
```
And here

ivs-cetmix commented on cetmix_tower_server/tests/test_reference_mixin.py:
```diff
@@ -92,3 +92,109 @@ def test_search_by_reference(self):
+            "Entries missing the required field should have a default reference.",
+        )
+
+    def test_handling_duplicate_ids(self):
```
And here

ivs-cetmix commented on cetmix_tower_server/tests/test_reference_mixin.py:
```diff
@@ -92,3 +92,109 @@ def test_search_by_reference(self):
+            "The second duplicate reference should have the correct suffix.",
+        )
+
+    def test_empty_vals_list_for_populate_references(self):
```
And here

dessanhemrayev commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -129,7 +128,8 @@
         if not self._context.get("reference_mixin_override"):
             for vals in vals_list:
-                vals["name"] = vals["name"].strip()
+                if vals.get("name"):
+                    vals["name"] = vals["name"].strip()
```
Thank you for pointing this out! After consideration, I realized that calling strip() on an empty string has no effect—it will remain empty. Therefore, we don't need to modify the condition to handle empty strings explicitly. The current condition if vals.get("name"): is sufficient, as it ensures strip() is only called when name has a meaningful, non-empty value.

This approach also skips any unnecessary processing on empty or None values, improving efficiency.

dessanhemrayev commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,84 @@ def _get_id_by_reference(self, reference):
         return records and records[0].id
+
+    @api.model
+    def _prepare_references(self, model, key_name, vals_list):
+        """
+        Prepare a dictionary of references for given model records.
+
+        This function extracts unique IDs from a list of dictionaries (vals_list)
+        based on a specified key (key_name), fetches the corresponding records
+        from the specified model, and returns a dictionary mapping record IDs to
+        their references.
+
+        Args:
+            model (str): The name of the model to fetch records from.
+            key_name (str): The key in the dictionaries of vals_list that contains
+                            the record IDs.
+            vals_list (list of dict): A list of dictionaries containing the values
+                                    to be processed.
+
+        Returns:
+            dict: A dictionary mapping record IDs to their references.
+        """
```
Thank you for pointing this out! I'll fix it

dessanhemrayev commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,84 @@
+            if record_id:
+                line_index_dict[record_id] += 1
+                line_index = line_index_dict[record_id]
+                vals["reference"] = f"{refs[record_id]}{suffix}_{line_index}"
```
Thank you for pointing this out! I'll fix it

dessanhemrayev commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,84 @@
+            if record_id:
+                line_index_dict[record_id] += 1
+                line_index = line_index_dict[record_id]
+                vals["reference"] = f"{refs[record_id]}{suffix}_{line_index}"
```
Thank you for pointing this out! I'll fix it

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,88 @@ def _get_id_by_reference(self, reference):
+            CxModel = self.env[model]
+        except KeyError as err:
+            raise ValueError(
+                _("Model '%s' does not exist. Please provide a valid model name.")
```
Please use correct translated strings implementation

ivs-cetmix commented:
/ocabot rebase

CetmixGitDrone commented:
Congratulations, PR rebased to [14.0-dev](https://github.com/cetmix/cetmix-tower/commits/14.0-dev).

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -10,6 +10,9 @@
 
 class CxTowerPlanLine(models.Model):
     _name = "cx.tower.plan.line"
+    _inherit = [
+        "cx.tower.reference.mixin",
+    ]
```
This is not relevant

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -9,12 +10,14 @@
-    variable_id = fields.Many2one(
-        string="Variable", comodel_name="cx.tower.variable", required=True
-    )
+    variable_id = fields.Many2one(string="Variable", comodel_name="cx.tower.variable")
```
Why do you remove "required" attribute here?

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_variable_value.py:
```diff
@@ -279,3 +282,11 @@ def _check_single_assignment(self):
                     )
                 )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # Use the common method from mixin
+        vals_list = self._populate_references(
+            "cx.tower.variable", "variable_id", vals_list, suffix="_variable"
+        )
+        return super().create(vals_list)
```
This method is used only in several model.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -242,3 +250,90 @@
+
+        # Fetch all line references in a single query
+        lines = CxModel.browse(line_ids)
+        return {line.id: line.reference for line in lines}
```
This suggestion is wrong. Current implementation is done according to the Odoo official documentation:
https://www.odoo.com/documentation/16.0/developer/howtos/translations.html#variables

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_reference_mixin.py:
```diff
@@ -186,10 +186,14 @@ def _get_copied_name(self):
 
         counter = 1
+        # Ensures that the generated copy name is unique by
+        # appending a counter until a unique name is found.
         while self.search_count([("name", "=", copy_name)]) > 0:
             counter += 1
             copy_name = _(
-                "%(name)s (copy %(number)s)", name=original_name, number=str(counter)
+                "%(name)s (copy %(number)s)",
+                name=original_name,
+                number=str(counter),
```
This suggestion is wrong. Current implementation is done according to the Odoo official documentation:
https://www.odoo.com/documentation/16.0/developer/howtos/translations.html#variables

ivs-cetmix reviewed (changes requested):
Please add the 'reference` field 'optional=hide' after the name field:
- for plan lines
![image](https://github.com/user-attachments/assets/d83f74f7-3222-4bb6-8691-2ce45c5c1ab0)

- for plan line actions
![image](https://github.com/user-attachments/assets/675ee2da-9438-46d6-bbc9-956b068452a6)

- for post run action variables
![image](https://github.com/user-attachments/assets/676fde37-dc42-4365-9890-f0766814ea67)

Aslo change 'variable_reference' field of the model 'cx.variable.value' model to "Variable Reference" (just remove the 'string=...' in the model)

Aldeigja reviewed (approved):
Functional review LGTM

ivs-cetmix reviewed (approved):
LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

ivs-cetmix commented:
/ocabot merge patch

CetmixGitDrone commented:
What a great day to merge this nice PR. Let's do it!
Prepared branch [14.0-dev-ocabot-merge-pr-129-by-ivs-cetmix-bump-patch](https://github.com/cetmix/cetmix-tower/commits/14.0-dev-ocabot-merge-pr-129-by-ivs-cetmix-bump-patch), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 6a992c06a71dcb5d751432f117d3d503376b6863. Thanks a lot for contributing to cetmix. ❤️
